h/static/scripts/base/controller.js

```diff
@@ -86,6 +106,95 @@ class Controller {
   forceUpdate() {
     this.update(this.state, this.state);
   }
+
+  /**
+   * Replace the HTML content of the element with an updated version from the
+   * server.
+   *
+   * Since this replaces the entire element, the current controller will not
+   * receive events for the replaced element. Instead the new controller for the
+   * replaced element is returned and can be used to transfer state from the
+   * existing controller.
+   *
+   * @param {string} html - The new markup for the element
+   * @return {Controller} - The new controller instance of the same type as this
+   *         controller on the element that replaces `this.element`
+   */
+  reload(html) {
+    if (!this.options.reload) {
```

Comment:
Wait, so this function is just a wrapper around a function which is passed in from outside the controller? That's a really odd inversion of control. Either the controller base class should know how to apply upgrades itself, and use that functionality here, or `reload` itself should be part of the `upgrade-elements` module, and not part of a controller.